File: Scores/Mapper/StudentMapper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Scores.Interfaces;
using Scores.Models;

namespace Scores
{
    public class StudentMapper : IStudentMapper
    {
        private readonly ISortingStrategyFactory _strategyFactory;
        private readonly ILogger<StudentMapper> _logger;

        public StudentMapper(ISortingStrategyFactory strategyFactory, ILogger<StudentMapper> logger)
        {
            _strategyFactory = strategyFactory ?? throw new ArgumentNullException(nameof(strategyFactory));
            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        }

        public Task<List<StudentSubject>> MapToStudentSubjectsAsync(List<StudentScoreRow> rows)
        {
            _logger.LogInformation("Mapping {Count} rows to student subjects", rows.Count);

            var studentSubjects = rows
                .GroupBy(r => new { r.StudentId, r.Name, r.Subject })
                .AsParallel()
                .Select(group => new StudentSubject
                {
                    StudentId = group.Key.StudentId,
                    StudentName = group.Key.Name,
                    Subject = group.Key.Subject,
                    Scores = group.Select(r => new StudentScore
                    {
                        LearningObjective = r.LearningObjective,
                        Score = r.Score
                    }).ToList()
                })
                .ToList();

            _logger.LogInformation("Mapped to {Count} student subjects", studentSubjects.Count);
            return Task.FromResult(studentSubjects);
        }

        public async Task<List<StudentSubject>> MapAndSortScoresAsync(List<StudentScoreRow> rows)
        {
            _logger.LogInformation("Starting mapping and sorting process for {Count} rows", rows.Count);

            var studentSubjects = await MapToStudentSubjectsAsync(rows);

            await Task.Run(() =>
            {
                System.Threading.Tasks.Parallel.ForEach(studentSubjects, student =>
                {
                    _logger.LogDebug("Sorting scores for student {StudentId} - {Subject}", student.StudentId, student.Subject);
                    var strategy = _strategyFactory.GetStrategy(student.Subject);
                    student.Scores = strategy.SortScores(student.Scores);
                });
            });

            _logger.LogInformation("Mapping and sorting completed successfully");
            return studentSubjects;
        }
    }
}

File: scores-php/src/Services/StudentMapper.php
<?php

namespace Scores\Services;

use Scores\Interfaces\IStudentMapper;
use Scores\Interfaces\ISortingStrategyFactory;
use Scores\Models\StudentScore;
use Scores\Models\StudentSubject;
use Psr\Log\LoggerInterface;

class StudentMapper implements IStudentMapper
{
    public function __construct(
        private ISortingStrategyFactory $strategyFactory,
        private LoggerInterface $logger
    ) {}
    
    public function mapToStudentSubjects(array $rows): array
    {
        $this->logger->info("Mapping " . count($rows) . " rows to student subjects");
        
        $grouped = [];
        
        foreach ($rows as $row) {
            $key = "{$row->studentId}_{$row->subject}";
            
            if (!isset($grouped[$key])) {
                $grouped[$key] = new StudentSubject(
                    studentId: $row->studentId,
                    studentName: $row->name,
                    subject: $row->subject,
                    scores: []
                );
            }
            
            $grouped[$key]->scores[] = new StudentScore(
                learningObjective: $row->learningObjective,
                score: $row->score
            );
        }
        
        $studentSubjects = array_values($grouped);
        
        $this->logger->info("Mapped to " . count($studentSubjects) . " student subjects");
        
        return $studentSubjects;
    }
    
    public function mapAndSortScores(array $rows): array
    {
        $this->logger->info("Starting mapping and sorting process for " . count($rows) . " rows");
        
        $studentSubjects = $this->mapToStudentSubjects($rows);
        
        foreach ($studentSubjects as $student) {
            $this->logger->debug("Sorting scores for student {$student->studentId} - {$student->subject}");
            $strategy = $this->strategyFactory->getStrategy($student->subject);
            $student->scores = $strategy->sortScores($student->scores);
        }
        
        $this->logger->info("Mapping and sorting completed successfully");
        
        return $studentSubjects;
    }
    
    public function mapAndSortScoresFromIterable(iterable $rows): array
    {
        $this->logger->info("Starting streaming mapping and sorting process");
        
        $grouped = [];
        $processedCount = 0;
        
        foreach ($rows as $chunk) {
            foreach ($chunk as $row) {
                $key = "{$row->studentId}_{$row->subject}";
                
                if (!isset($grouped[$key])) {
                    $grouped[$key] = new StudentSubject(
                        studentId: $row->studentId,
                        studentName: $row->name,
                        subject: $row->subject,
                        scores: []
                    );
                }
                
                $grouped[$key]->scores[] = new StudentScore(
                    learningObjective: $row->learningObjective,
                    score: $row->score
                );
                
                $processedCount++;
            }
            
            $this->logger->debug("Processed chunk: {$processedCount} total rows processed so far");
            
            unset($chunk);
        }
        
        $this->logger->info("Mapped {$processedCount} rows to " . count($grouped) . " student subjects");
        
        // Sort scores for each student subject
        $studentSubjects = array_values($grouped);
        
        foreach ($studentSubjects as $student) {
            $this->logger->debug("Sorting scores for student {$student->studentId} - {$student->subject}");
            $strategy = $this->strategyFactory->getStrategy($student->subject);
            $student->scores = $strategy->sortScores($student->scores);
        }
        
        $this->logger->info("Streaming mapping and sorting completed successfully");
        
        return $studentSubjects;
    }
}

